### DISCUSSION TITLE
Slow execution of jobs using Git sync

### DISCUSSION THREAD

**Discussion by JanHerca** [requarks/wiki#5562] [Help / Questions]
Using Git to sync a GitHub project in bi-directional mode (every 5 min) I am finding in Docker logs a very slow execution of jobs. I am adding several thousands of pages and each one can easily take 1-2 minutes per page, which means I need days for each GitHub commit.

Any help on how to do faster loads? Is there a config file to increase number of concurrent jobs or to speed them up?

OS (Docker Linux in Digital Ocean on Ubuntu)
CPU cores 2
Total RAM 1.95 GB
Node.js 14.17.6
PostgreSQL 11.8
Wiki.js 2.5.219

**Comment by JanHerca** [comment]
OK. I can give my own answer after more research. Seems that what slows down each Git storage job is the rebuild of page tree when a page is marked as new, when it is a new created. As my website is having a good amount of pages (40k+), and rebuilding page tree rebuilds the entire table of website tree, that is the reason of the long time it takes that step in the job. Rendering is very fast (2 secs) but rebuilding the tree can take my site 2mins.

The rebuilding of the tree should be optimized in future versions, because if your site is going to have 40k+ pages each time a user creates a new page from the web it must await about 1-2 min (depending on server) to finally see the page created. If a user is going to create many pages from the web it is very annoying to have to wait so much in each new page created.

**Reply by fumbles** [reply]
We’re at 60k pages and it’s a struggle when we have a tool that creates two pages at a time (file, folder and second file). Graphql has been timing out waiting for the rebuild process.

**Comment by JanHerca** [comment]
More info about limitations of Wiki.js: [https://github.com/requarks/wiki/discussions/5602](https://github.com/requarks/wiki/discussions/5602)

**Comment by rhvitfeldt** [comment]
We are in the same situation. We are bulk deleting and creating 15 000 pages with GraphQL api. Every single delete takes approx. 12 seconds. The bottleneck seems to be the rebuild of page tree. We have implemented a queue that will wait with next page delete until wiki responds OK to the previous one. Deleting/creating all pages will therefore take 50 hours.

Can we somehow increase the speed of page tree rebuild? Or some other way make this process faster?

Thank you!

@NGPixel

**Comment by NGPixel** [comment]
This is being investigated and the impact should be minimized in the next patch update.

Note that the upcoming 3.0 version use a different tree system that doesn't need to be rebuilt every time.

**Reply by DavidOhsie** [reply]
@NGPixel Is there a way to temporary defer page tree rebuild in the system until after a large import?  I'm trying to import data from another wiki and I don't need the page table rebuild during that process. I can wait until the end.
